jet requires JCache - 0.3.2-SNASHOT

last updated snapshot requires jcache `20170328042018`

```
Exception in thread "main" java.lang.NoClassDefFoundError: javax/cache/Cache$Entry
	at com.hazelcast.jet.impl.execution.init.JetSerializerHook$CacheEntry.getSerializationType(JetSerializerHook.java:148)
	at com.hazelcast.internal.serialization.impl.SerializerHookLoader.load(SerializerHookLoader.java:63)
	at com.hazelcast.internal.serialization.impl.SerializerHookLoader.<init>(SerializerHookLoader.java:55)
	at com.hazelcast.internal.serialization.impl.DefaultSerializationServiceBuilder.registerSerializerHooks(DefaultSerializationServiceBuilder.java:282)
	at com.hazelcast.internal.serialization.impl.DefaultSerializationServiceBuilder.build(DefaultSerializationServiceBuilder.java:223)
	at com.hazelcast.internal.serialization.impl.DefaultSerializationServiceBuilder.build(DefaultSerializationServiceBuilder.java:49)
	at com.hazelcast.instance.DefaultNodeExtension.createSerializationService(DefaultNodeExtension.java:151)
	at com.hazelcast.instance.Node.<init>(Node.java:201)
	at com.hazelcast.instance.HazelcastInstanceImpl.createNode(HazelcastInstanceImpl.java:163)
	at com.hazelcast.instance.HazelcastInstanceImpl.<init>(HazelcastInstanceImpl.java:130)
	at com.hazelcast.instance.HazelcastInstanceFactory.constructHazelcastInstance(HazelcastInstanceFactory.java:218)
	at com.hazelcast.instance.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:176)
	at com.hazelcast.instance.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:126)
	at com.hazelcast.core.Hazelcast.newHazelcastInstance(Hazelcast.java:58)
	at com.hazelcast.jet.Jet.newJetInstance(Jet.java:49)
	at com.hazelcast.jet.Jet.newJetInstance(Jet.java:58)
	at com.hazelcast.techops.jet.training.essentials.MyJetInstance.main(MyJetInstance.java:19)
Caused by: java.lang.ClassNotFoundException: javax.cache.Cache$Entry
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 17 more
```

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/ReadWithPartitionIteratorP.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/ReadWithPartitionIteratorP.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.jet.impl.connector;
 
+import com.hazelcast.cache.impl.CacheEntry;
 import com.hazelcast.cache.impl.CacheProxy;
 import com.hazelcast.client.cache.impl.ClientCacheProxy;
 import com.hazelcast.client.config.ClientConfig;
@@ -35,6 +36,7 @@ import com.hazelcast.map.impl.proxy.MapProxyImpl;
 import com.hazelcast.nio.Address;
 
 import javax.annotation.Nonnull;
+import java.util.AbstractMap;
 import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
@@ -102,8 +104,8 @@ public final class ReadWithPartitionIteratorP<T> extends AbstractProcessor {
 
     public static <T> ProcessorMetaSupplier readCache(String cacheName, int fetchSize) {
         return new LocalClusterMetaSupplier<T>(
-                instance -> partition -> ((CacheProxy) instance.getCacheManager().getCache(cacheName))
-                        .iterator(fetchSize, partition, PREFETCH_VALUES));
+                instance -> partition -> wrap(((CacheProxy) instance.getCacheManager().getCache(cacheName))
+                        .iterator(fetchSize, partition, PREFETCH_VALUES)));
     }
 
     public static ProcessorMetaSupplier readCache(String cacheName, ClientConfig clientConfig) {
@@ -112,8 +114,8 @@ public final class ReadWithPartitionIteratorP<T> extends AbstractProcessor {
 
     public static <T> ProcessorMetaSupplier readCache(String cacheName, int fetchSize, ClientConfig clientConfig) {
         return new RemoteClusterMetaSupplier<T>(clientConfig,
-                instance -> partition -> ((ClientCacheProxy) instance.getCacheManager().getCache(cacheName))
-                        .iterator(fetchSize, partition, PREFETCH_VALUES));
+                instance -> partition -> wrap(((ClientCacheProxy) instance.getCacheManager().getCache(cacheName))
+                        .iterator(fetchSize, partition, PREFETCH_VALUES)));
     }
 
     @Override
@@ -269,5 +271,29 @@ public final class ReadWithPartitionIteratorP<T> extends AbstractProcessor {
             return getProcessors(count, ownedPartitions, partitionToIterator);
         }
 
+    }
+
+    private static Iterator wrap(Iterator iterator) {
+        return new IteratorWrapper(iterator);
+    }
+
+    private static class IteratorWrapper implements Iterator<Map.Entry> {
+
+        private final Iterator iterator;
+
+        IteratorWrapper(Iterator iterator) {
+            this.iterator = iterator;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return iterator.hasNext();
+        }
+
+        @Override
+        public Map.Entry next() {
+            CacheEntry next = (CacheEntry) iterator.next();
+            return new AbstractMap.SimpleImmutableEntry(next.getKey(), next.getValue());
+        }
     }
 }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/JetSerializerHook.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/JetSerializerHook.java
@@ -23,7 +23,6 @@ import com.hazelcast.nio.serialization.Serializer;
 import com.hazelcast.nio.serialization.SerializerHook;
 import com.hazelcast.nio.serialization.StreamSerializer;
 
-import javax.cache.Cache;
 import java.io.IOException;
 import java.util.Map;
 import java.util.Map.Entry;
@@ -131,45 +130,6 @@ public final class JetSerializerHook {
                 @Override
                 public Entry read(ObjectDataInput in) throws IOException {
                     return entry(in.readObject(), in.readObject());
-                }
-            };
-        }
-
-        @Override
-        public boolean isOverwritable() {
-            return true;
-        }
-    }
-
-    public static final class CacheEntry implements SerializerHook<Cache.Entry> {
-
-        @Override
-        public Class<Cache.Entry> getSerializationType() {
-            return Cache.Entry.class;
-        }
-
-        @Override
-        public Serializer createSerializer() {
-            return new StreamSerializer<Cache.Entry>() {
-                @Override
-                public int getTypeId() {
-                    return CACHE_ENTRY;
-                }
-
-                @Override
-                public void destroy() {
-
-                }
-
-                @Override
-                public void write(ObjectDataOutput out, Cache.Entry object) throws IOException {
-                    out.writeObject(object.getKey());
-                    out.writeObject(object.getValue());
-                }
-
-                @Override
-                public Cache.Entry read(ObjectDataInput in) throws IOException {
-                    return new com.hazelcast.cache.impl.CacheEntry<>(in.readObject(), in.readObject());
                 }
             };
         }

--- a/hazelcast-jet-core/src/main/resources/META-INF/services/com.hazelcast.SerializerHook
+++ b/hazelcast-jet-core/src/main/resources/META-INF/services/com.hazelcast.SerializerHook
@@ -1,4 +1,3 @@
 com.hazelcast.jet.impl.execution.init.JetSerializerHook$MapEntry
 com.hazelcast.jet.impl.execution.init.JetSerializerHook$ObjectArray
 com.hazelcast.jet.impl.execution.init.CustomClassLoadedObject$Hook
-com.hazelcast.jet.impl.execution.init.JetSerializerHook$CacheEntry

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/HazelcastConnectorIntegrationTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/HazelcastConnectorIntegrationTest.java
@@ -30,18 +30,15 @@ import org.junit.Before;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 
-import javax.cache.Cache;
 import java.util.concurrent.ExecutionException;
 
 import static com.hazelcast.jet.Edge.between;
-import static com.hazelcast.jet.Processors.map;
 import static com.hazelcast.jet.Processors.readCache;
 import static com.hazelcast.jet.Processors.readList;
 import static com.hazelcast.jet.Processors.readMap;
 import static com.hazelcast.jet.Processors.writeCache;
 import static com.hazelcast.jet.Processors.writeList;
 import static com.hazelcast.jet.Processors.writeMap;
-import static com.hazelcast.jet.Util.entry;
 import static java.util.stream.Collectors.toList;
 import static java.util.stream.IntStream.range;
 import static org.junit.Assert.assertEquals;
@@ -90,11 +87,9 @@ public class HazelcastConnectorIntegrationTest extends JetTestSupport {
 
         DAG dag = new DAG();
         Vertex source = dag.newVertex("source", readCache(sourceName));
-        Vertex mapper = dag.newVertex("mapper", map((Cache.Entry e) -> entry(e.getKey(), e.getValue())));
         Vertex sink = dag.newVertex("sink", writeCache(sinkName));
 
-        dag.edge(between(source, mapper));
-        dag.edge(between(mapper, sink));
+        dag.edge(between(source, sink));
 
         jetInstance.newJob(dag).execute().get();
 

